### ISSUE TITLE
Workshop - Haxe tutorial for beginners

### ISSUE THREAD

**Issue by andyli** [HaxeSummit2017/organization#15]
I haven't decided yet, but I would like to have some discussion on the idea of giving a beginner-level Haxe workshop during the summit.

I and @kevinresol held an [introductory Haxe workshop](https://github.com/kevinresol/hkoscon2017-haxe-game) in the previous HKOSCon. It attracted quite a lot of ppl to participate. Although, the content we prepared was a bit too much so we skipped a lot of details we initially wanted to cover.

I'm thinking, maybe it would be good to reused the material and host a similar workshop in the summit, which targets people who are interested in Haxe, but haven't really started using yet. So, the summit would not be limited to only existing Haxe users, who can now bring their friends (or significant other) to learn a bit of what Haxe is about, by buying a 1-day ticket to attend the workshop.

While I and Kevin will be the main instructors, the other Haxe devs/users will help the beginners to follow the workshop.

The workshop would also give me some motivation to finish the Haxe MOOC course I proposed a long time ago :P

**Comment by andyli** [comment]
Let's have a little poll here for me to estimate:
 * Give me a :+1: if you're a Haxe user, and would (try to) bring a (non-Haxe user) friend to this workshop.
 * Give me a :heart: if you're a Haxe user who would help the beginners in the workshop, though you couldn't bring any friend to attend.
 * Give me a :smile: if you're yourself a Haxe beginner or non-Haxe user, and would like to learn Haxe in the workshop.

**Comment by kevinresol** [comment]
無啦啦拉我落水?! :joy:
(sorry for non-cantonese speakers, I just dunno how to express that in english)

**Comment by andyli** [comment]
This is just a plan for now, no pressure...
But I know you enjoy teaching Haxe 😈

**Comment by back2dos** [comment]
This is absolutely great. I think there's a big target audience in Amsterdam that we could invite (we have a campus nearby and our offices are in a coworking space filled with techies). If you need someone to confuse people, count me in :D

@kevinresol you can do it! ;)

**Comment by andyli** [comment]
Ok, let's do it :D

**Comment by andyli** [comment]
Hey, I think it's time to get things rolling!

My initial idea is to base on the material we used in https://github.com/kevinresol/hkoscon2017-haxe-game.
Since we overran in the 2 hour time slot last time, I expect 3 hour would be better this time.
Over 3 hour could be too tiring I guess.

Let me try convert the existing material into an assignment format, more like a step-by-step guide to go from nothing to building a complete game.

@back2dos 
Do you want to hold a 2nd edition of the previous macros workshop? I am thinking, it would be nice if you can handle the macros part, separated from the intro workshop. I and @kevinresol tried to include a small macros example, but we think there is too much detail and too advanced to cover in an intro workshop.

**Comment by back2dos** [comment]
Hmm, well, a slight problem is that 3h leaves a 1h slot, which is kinda hard to fit a workshop into. My suggestion would be to plan for 4h with sufficient breaks. But if you're reasonably sure that such a duration would be pointless, I'll figure something out.

----

As for the macro workshop, I'd rather not. I think it's more useful for someone else to talk on the subject to get a different perspective that'll complement last year's material. If anyone feels motivated, I'll of course be glad to help in any way I can ;)

**Comment by nadako** [comment]
Regarding macros I feel like we need a "best practices" talk/workshop for seasoned Haxe users. Macros, with all their power often turn out being slow and unmaintainable mess, so some advices regarding clean macro code (e.g. using reification, #if fencing, etc.), performance (encoding/decoding and whatnot), cache server nuances (module dependencies, etc) and general tips&tricks would really help a lot. The sad news is that I think @ncannasse is the only human having all this knowledge :) Hope I'm wrong.

**Comment by andyli** [comment]
Umm... I think 4h with sufficient breaks is fine. It can be arranged as 6 x 30min sections with 10min breaks in between. The number of sections matches exactly how we've planed the workshop initially. What do you think, @kevinresol?

**Comment by kevinresol** [comment]
I doubt if any one really wish to sit there for 4 hours to build a dumb game.
Personally I dislike workshops (because the pace just never fit me), and I never really attended one in full. So I am not really sure how to run one...

**Comment by andyli** [comment]
It's not really for building a game, but for learning a language. I believe 4 hours is rather short for either one...

Anyway, do you have any suggestion?

**Comment by kevinresol** [comment]
Maybe we could breakdown the whole thing into sections and provide reproducible code at each stage.
So people who come in late could still continue from some point.

**Comment by andyli** [comment]
Yes, I think that's what we should have done. Since we've more time now, I think we can add some regular beginner exercisers in the first section.

e.g.
Write a function that prints
```
****
***
**
*
```

Write a function that computes the distance between two points - which will be used when building the game.

**Comment by PippoApps** [comment]
About to start a large game in browser without Flash Player (first time I do accept this kind of contract offer)... I used to code in pure JS, even very complex stuff, but that's not suitable for a cross-platform large online game (unless days were made of 64 hours). Any Haxe enthusiast wishes to summarise how Haxe exports for browser and devices? I.e., in browser, does it export for canvas? Etc. Some kind soul here might spare me a day of research :) <3

**Comment by back2dos** [comment]
Hey @PippoApps, this is hardly the right place to discuss such questions. The short answer is: it depends. Haxe merely compiles to the language you want. If that's JS you can either draw to canvas or manipulate the DOM. You can use OpenFl which basically gives you the flash API across various platforms. On browser that'll be canvas. For more details, please ask in the Haxe Google group, the OpenFl forums or stackoverflow ;)

**Comment by PippoApps** [comment]
Ah you are right hardly the right place, I am sorry. Thank you for the answer!
